README
This script is designed to query Zenoss instances, and report back on Events.

Requirements:
* Python
* PyYAML (http://pyyaml.org)
  * If installing on Linux Mint, run 'sudo apt-get install python-yaml'
    May also be available in Ubuntu.


CHANGELOG
0.1
* Initial Release

0.2
* Started adding version numbers.
* Added timeout value to Zenoss queries.  Zenoss authentication is set to twice the timeout value.
* Added new version that provides minimal output, using the curses library for screen updates.
* Removed all commandline options, and added new commandline option to specify optional configuration file.
  The configuration file can be used to override any default options.  A sample config file is also
  provided, as well.
* Added shell scripts to run both versions with the config file named 'config'.
* Checked in the .svnignore file, along with a shell script to apply updates to the ignore file.


api.py
# Zenoss-4.x JSON API Example (python)
#
# To quickly explore, execute 'python -i api_example.py'
#
# >>> z = ZAPI()
# >>> events = z.get_events()
# etc.

import json
import socket
import urllib
import urllib2
import ssl

ROUTERS = { 'MessagingRouter': 'messaging',
            'EventsRouter': 'evconsole',
            'ProcessRouter': 'process',
            'ServiceRouter': 'service',
            'DeviceRouter': 'device',
            'NetworkRouter': 'network',
            'TemplateRouter': 'template',
            'DetailNavRouter': 'detailnav',
            'ReportRouter': 'report',
            'MibRouter': 'mib',
            'ZenPackRouter': 'zenpack' }

# Let's set a timeout value of 5 seconds
QUERY_TIMEOUT = 5
## Let's set the default timeout (in seconds) for all socket connections
#socket.setdefaulttimeout(5)

class ZAPI():
    targetInstance='http://localhost:8080'
    def __init__(self, debug=False, targetInstance=targetInstance, zenossUser='admin', zenossPassword='zenoss'):
        """
        Initialize the API connection, log in, and store authentication cookie
        """
        self.targetInstance=targetInstance
        # Use the HTTPCookieProcessor as urllib2 does not save cookies by default
        handler=urllib2.HTTPSHandler(context=ssl._create_unverified_context())
        self.urlOpener = urllib2.build_opener(urllib2.HTTPCookieProcessor(),handler)
        if debug: self.urlOpener.add_handler(urllib2.HTTPHandler(debuglevel=1))
        self.reqCount = 1

        # Contruct POST params and submit login.
        loginParams = urllib.urlencode(dict(
                        __ac_name = zenossUser,
                        __ac_password = zenossPassword,
                        submitted = 'true',
                        came_from = targetInstance + '/zport/dmd'))
        self.urlOpener.open(targetInstance + '/zport/acl_users/cookieAuthHelper/login',
                            loginParams, QUERY_TIMEOUT*2)

    def _router_request(self, router, method, data=[]):
        if router not in ROUTERS:
            raise Exception('Router "' + router + '" not available.')

        # Contruct a standard URL request for API calls
        req = urllib2.Request(self.targetInstance + '/zport/dmd/' +
                              ROUTERS[router] + '_router')

        # NOTE: Content-type MUST be set to 'application/json' for these requests
        req.add_header('Content-type', 'application/json; charset=utf-8')

        # Convert the request parameters into JSON
        reqData = json.dumps([dict(
                    action=router,
                    method=method,
                    data=data,
                    type='rpc',
                    tid=self.reqCount)])

        # Increment the request count ('tid'). More important if sending multiple
        # calls in a single request
        self.reqCount += 1

        # Submit the request and convert the returned JSON to objects
        return json.loads(self.urlOpener.open(req, reqData, QUERY_TIMEOUT).read())

    def get_devices(self, deviceClass='/zport/dmd/Devices'):
        return self._router_request('DeviceRouter', 'getDevices',
                                    data=[{'uid': deviceClass,
                                           'params': {} }])['result']

    def close_events(self, evids=None):
        return self._router_request('EventsRouter', 'close', data=[{'evids': evids,'params': {'severiy':[5,4,3,2]} }])['result']

    def get_events2(self, device=None, component=None, deviceClass='network', eventClass=None):
        data = dict(start=0, limit=100, dir='DESC', sort='severity')
        data['params'] = dict(severity=[5], eventState=[0],DeviceClass=deviceClass)

        if device: data['params']['device'] = device
        if component: data['params']['component'] = component
        if eventClass: data['params']['eventClass'] = eventClass

        return self._router_request('EventsRouter', 'query', [data])['result']

    def get_events(self, device=None, component=None, deviceClass='network', eventClass=None, severity=[5,4,3,2], eventState=[0,1], prodState=[1000]):
        data = dict(start=0, limit=100, dir='DESC', sort='severity')
        data['params'] = dict(severity=severity, eventState=eventState, DeviceClass=deviceClass)

        if device: data['params']['device'] = device
        if component: data['params']['component'] = component
        if prodState: data['params']['prodState'] = prodState

        return self._router_request('EventsRouter', 'query', [data])['result']

    def get_events_uid(self, device=None, component=None, uid="/zport/dmd/Groups/Office", eventClass=None, severity=[5,4,3,2], eventState=[0,1], prodState=[1000]):
        data = dict(start=0, limit=100, dir='DESC', sort='severity', uid=uid)
        data['params'] = dict(severity=severity, eventState=eventState)

        if device: data['params']['device'] = device
        if component: data['params']['component'] = component
        if prodState: data['params']['prodState'] = prodState
        return self._router_request('EventsRouter', 'query', [data])['result']


    def get_events_gen(self):
        data = dict(dir='DESC', sort='lastTime')
        data['params'] = dict(eventState=[])

        return self._router_request('EventsRouter', 'queryGenerator', [data])['result']

    def add_device(self, deviceName, deviceClass):
        data = dict(deviceName=deviceName, deviceClass=deviceClass)
        return self._router_request('DeviceRouter', 'addDevice', [data])

    def get_device_components(self, uid):
        if uid:
           return self._router_request('DeviceRouter', 'getComponents', data=[{'uid':uid}])


    def delete_device(self, uids):
        if uids:
            params={'productionState':['1000','500','400','300', '-1']}
            return self._router_request('DeviceRouter', 'removeDevices', data=[{'sort':'name', 'dir':'ASC', 'uids':uids, 'hashcheck':'0', 'action':'delete', 'deleteEvents':'true', 'deletePerf':'true', 'params': params }])['result']


    def create_event_on_device(self, device, severity, summary):
        if severity not in ('Critical', 'Error', 'Warning', 'Info', 'Debug', 'Clear'):
            raise Exception('Severity "' + severity +'" is not valid.')

        data = dict(device=device, summary=summary, severity=severity,
                    component='', evclasskey='', evclass='')
        return self._router_request('EventsRouter', 'add_event', [data])


beepScript.py
#!/usr/bin/env python

import __main__
import curses
import logging
import os
import sys
import traceback
import time
import yaml
from api import ZAPI
from optparse import OptionParser
from random import randint
from random import random
import socket
from time import gmtime, strftime
import urllib2
if sys.platform.lower().startswith('linux'):
    import glib
    import gobject
    import thread
    from sound import playSound

# Setup app_path
app_path = os.path.dirname(os.path.abspath(getattr(__main__,'__file__','__main__.py')))

VERSION = 0.2

TIMESTAMP_ROW = 0
NAME_ROW = 1
ALERTS_ROW = 2

severities = [5]
event_states = [0]
production_states = [1000]
device_classes = '/Network*'
cycle_time = 15
username = 'username'
password = 'password'
sound_file = app_path + '/console-beep.mp3'
instances = [{"name": "Localhost", "short_name": "Lo", "target_instance": "http://localhost:8080"}]


# Option Parser
parser = OptionParser()
parser.add_option('-c', '--config-file', dest='config_file', help='Configuration file')
(options, args) = parser.parse_args()

# Use configuration file specified on commandline, if given
if options.config_file:
    config_file_options = yaml.load(open(options.config_file))
    if config_file_options.has_key('severities'): severities = config_file_options['severities']
    if config_file_options.has_key('event_states'): event_states = config_file_options['event_states']
    if config_file_options.has_key('production_states'): production_states = config_file_options['production_states']
    if config_file_options.has_key('device_classes'): device_classes = config_file_options['device_classes']
    if config_file_options.has_key('cycle_time'): cycle_time = config_file_options['cycle_time']
    if config_file_options.has_key('username'): username = config_file_options['username']
    if config_file_options.has_key('password'): password = config_file_options['password']
    if config_file_options.has_key('sound_file'): sound_file = app_path + "/" + config_file_options['sound_file']
    if config_file_options.has_key('instances'): instances = config_file_options['instances']


# Setup Logging and SysLogger
levels = {
    'debug':    logging.DEBUG,
    'info':     logging.INFO,
    'warning':  logging.WARNING,
    'error':    logging.ERROR,
    'critical': logging.CRITICAL
}

logName = app_path + "/logs/zenoss_beeps.log"
level = levels.get('debug', logging.NOTSET)
logging.basicConfig(filename=logName, level=level)

ZenBeep = logging.getLogger('ZenBeep')
ZenBeep.setLevel(logging.DEBUG)

class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'

    def disable(self):
        self.HEADER = ''
        self.OKBLUE = ''
        self.OKGREEN = ''
        self.WARNING = ''
        self.FAIL = ''
        self.ENDC = ''

# Method for Playback
def play_sound(sound_file):
    if sys.platform.lower().startswith('linux'):
        uri = 'file://' + sound_file
        global loop
        loop = glib.MainLoop()
        playSoundClass = playSound(uri, loop=loop)
        thread.start_new_thread(playSoundClass.start, ())
        gobject.threads_init()
        loop.run()
    elif sys.platform.lower().startswith('darwin'):
        os.system("afplay " + sound_file)

def timestamp():
    return strftime("%a, %d %b %Y %H:%M:%S", gmtime())

def timestamp_log():
    return strftime("%Y-%m-%d %H:%M:%S", gmtime())

def query(instance, severities, event_states, production_states, device_class):
    ZenBeep.info("\t%s: Querying %s" % (timestamp_log(), instance['short_name']))
    
    target_instance = instance['target_instance']
    zapi = ZAPI(debug = False, targetInstance = target_instance, zenossUser = username, zenossPassword = password)
    query_result = zapi.get_events(severity = severities, eventState = event_states, prodState = production_states, deviceClass = device_class)
    event_count = int(query_result['totalCount'])
    
    return event_count

def main():
    starting_column = 0
    # Setup some extra attributes that we need
    for instance in instances:
        instance['starting_column'] = starting_column
        instance['alert_line'] = False
        starting_column += len(instance['short_name']) + 2

    while 1:
        ZenBeep.info("\t%s: #### Starting new query cycle" % timestamp_log())
        os.system('clear')

        for instance in instances:
            num_events = 0
            
            # Run the query
            try:
                num_events = query(instance, severities, event_states, production_states, device_classes)
            except (urllib2.URLError, socket.timeout), e:
                ZenBeep.critical("\t%s: Query problem with %s (%s)" % (timestamp_log(), instance['short_name'], e))
                num_events = -1

            # Now lety's update
            if num_events > 0:
                print "%s%s events need to be ACK in %s at %s%s" % (bcolors.FAIL, str(num_events), instance['short_name'], timestamp(), bcolors.ENDC)
                play_sound(sound_file)
                ZenBeep.info("\t%s: %s events in %s" % (timestamp_log(), str(num_events), instance['short_name']))
            elif num_events < 0:
                print "%sProblem with query against %s at %s%s" % (bcolors.WARNING, instance['short_name'], timestamp(), bcolors.ENDC)
            else:
                print "%sAll events in %s are ACK at %s%s" % (bcolors.OKGREEN, instance['short_name'], timestamp(), bcolors.ENDC)
            
        time.sleep(cycle_time)

try:
    if len(instances) == 0:
       print "No targets specified"
       ZenBeep.info("No targets specified")
       sys.exit(0)

    if __name__=='__main__':
        main()

except (KeyboardInterrupt, SystemExit):
    sys.exit(0)


beepScript.minimal.py
#!/usr/bin/env python

import __main__
import curses
import logging
import os
import sys
import traceback
import time
import yaml
from api import ZAPI
from optparse import OptionParser
from random import randint
from random import random
import socket
from time import gmtime, strftime
import urllib2
if sys.platform.lower().startswith('linux'):
    import glib
    import gobject
    import thread
    from sound import playSound

# Setup app_path
app_path = os.path.dirname(os.path.abspath(getattr(__main__,'__file__','__main__.py')))

VERSION = 0.2

TIMESTAMP_ROW = 0
NAME_ROW = 1
ALERTS_ROW = 2

severities = [5]
event_states = [0]
production_states = [1000]
device_class = 'network'
cycle_time = 15
username = 'username'
password = 'password'
sound_file = app_path + '/console-beep.mp3'
instances = [{"name": "Localhost", "short_name": "Lo", "target_instance": "http://localhost:8080"}]


# Option Parser
parser = OptionParser()
parser.add_option('-c', '--config-file', dest='config_file', help='Configuration file')
(options, args) = parser.parse_args()

# Use configuration file specified on commandline, if given
if options.config_file:
    config_file_options = yaml.load(open(options.config_file))
    if config_file_options.has_key('severities'): severities = config_file_options['severities']
    if config_file_options.has_key('event_states'): event_states = config_file_options['event_states']
    if config_file_options.has_key('production_states'): production_states = config_file_options['production_states']
    if config_file_options.has_key('device_classes'): device_classes = config_file_options['device_classes']
    if config_file_options.has_key('cycle_time'): cycle_time = config_file_options['cycle_time']
    if config_file_options.has_key('username'): username = config_file_options['username']
    if config_file_options.has_key('password'): password = config_file_options['password']
    if config_file_options.has_key('sound_file'): sound_file = app_path + "/" + config_file_options['sound_file']
    if config_file_options.has_key('instances'): instances = config_file_options['instances']

# Setup app_path
app_path = os.path.dirname(os.path.abspath(getattr(__main__,'__file__','__main__.py')))

# Setup Logging and SysLogger
levels = {
    'debug':    logging.DEBUG,
    'info':     logging.INFO,
    'warning':  logging.WARNING,
    'error':    logging.ERROR,
    'critical': logging.CRITICAL
}

logName = app_path + "/logs/zenoss_beeps.log"
level = levels.get('debug', logging.NOTSET)
logging.basicConfig(filename=logName, level=level)

ZenBeep = logging.getLogger('ZenBeep')
ZenBeep.setLevel(logging.DEBUG)

# Method for Playback
def play_sound(sound_file):
    if sys.platform.lower().startswith('linux'):
        uri = 'file://' + sound_file
        global loop
        loop = glib.MainLoop()
        playSoundClass = playSound(uri, loop=loop)
        thread.start_new_thread(playSoundClass.start, ())
        gobject.threads_init()
        loop.run()
    elif sys.platform.lower().startswith('darwin'):
        os.system("afplay " + sound_file)

def timestamp():
    return strftime("%a, %d %b %Y %H:%M:%S", gmtime())

def timestamp_log():
    return strftime("%Y-%m-%d %H:%M:%S", gmtime())

def update_alerts(screen, instance, num_alerts):
    screen.addstr(NAME_ROW, instance['starting_column'], instance['short_name'], curses.color_pair(1))
    screen.addstr(ALERTS_ROW, instance['starting_column'], str(num_alerts), curses.color_pair(1))

def update_normal(screen, instance):
    screen.addstr(NAME_ROW, instance['starting_column'], instance['short_name'])

def update_ok(screen, instance, num_alerts):
    screen.addstr(NAME_ROW, instance['starting_column'], instance['short_name'], curses.color_pair(3))
    screen.addstr(ALERTS_ROW, instance['starting_column'], str(num_alerts), curses.color_pair(3))

def update_query_errors(screen, instance):
    screen.addstr(NAME_ROW, instance['starting_column'], instance['short_name'], curses.color_pair(2))

def update_timestamp(screen):
#    nowString = strftime("%a, %d %b %Y %H:%M:%S", gmtime())
    screen.addstr(TIMESTAMP_ROW, 0, timestamp())

def query(instance, severities, event_states, production_states, device_class):
    ZenBeep.info("\t%s: Querying %s" % (timestamp_log(), instance['short_name']))

    severities =[5]
    eventState =[0]

    target_instance = instance['target_instance']
#     print severities
#     print event_states
#     print production_states
#     print device_class
    zapi = ZAPI(debug = True, targetInstance = target_instance, zenossUser = username, zenossPassword = password)
    query_result = zapi.get_events(severity = severities, eventState = event_states, prodState = production_states, deviceClass = device_class)
    #query_result = zapi.get_events2()
    ZenBeep.info(query_result)
    event_count = int(query_result['totalCount'])

    return event_count

def main(screen):
    starting_column = 0
    # Setup some extra attributes that we need
    for instance in instances:
        instance['starting_column'] = starting_column
        instance['alert_line'] = False
        starting_column += len(instance['short_name']) + 2

    curses.use_default_colors()
    curses.init_pair(1, curses.COLOR_RED, -1)
    curses.init_pair(2, curses.COLOR_YELLOW, -1)
    curses.init_pair(3, curses.COLOR_GREEN, -1)
    max_y,max_x = screen.getmaxyx()

    curses.curs_set(0)
    while 1:
        ZenBeep.info("\t%s: #### Starting new query cycle" % timestamp_log())

        update_timestamp(screen)
        screen.refresh

        for instance in instances:
            update_normal(screen, instance)
            screen.refresh()
            num_events = 0

            # Run the query
            try:
                num_events = query(instance, severities, event_states, production_states, device_class)
            except (urllib2.URLError, socket.timeout), e:
                ZenBeep.critical("\t%s: Query problem with %s (%s)" % (timestamp_log(), instance['short_name'], e))
                num_events = -1

            # Let's clear out the previous alert count
            screen.addstr(ALERTS_ROW, instance['starting_column'], ' '*(len(instance['short_name'])+1))

            # Now lety's update
            if num_events > 0:
                update_alerts(screen, instance, num_events)
                screen.refresh()
                #play_sound(sound_file)
                ZenBeep.info("\t%s: %s events in %s" % (timestamp_log(), str(num_events), instance['short_name']))
            elif num_events < 0:
                update_query_errors(screen, instance)
                screen.refresh()
            else:
                update_ok(screen, instance, num_events)
                screen.refresh()

        time.sleep(cycle_time)

try:
    if len(instances) == 0:
       print "No targets specified"
       ZenBeep.info("No targets specified")
       sys.exit(0)

    if __name__=='__main__':
        curses.wrapper(main)

except (KeyboardInterrupt, SystemExit):
    sys.exit(0)
